File: Packages/TypeEngine/Sources/TypeEngine/SpatialModel.swift
import Foundation

/// Keyboard-adjacency likelihood model for the Icelandic layout.
///
/// Key centers live on a unit grid (1.0 = one key width) with iOS-style row
/// stagger. P(typed char | intended char) follows a Gaussian over the distance
/// between key centers: exp(-d² / 2σ²), σ ≈ 0.7 key widths. The corrector
/// consumes -log of that as a substitution cost; insertions, deletions and
/// transpositions carry tuned constant costs. All costs are in nats.
public struct SpatialModel: Sendable {

    public struct Costs: Sendable {
        /// User typed an extra character that was not intended.
        public var insertion: Double = 4.0
        /// User omitted a character they intended.
        public var deletion: Double = 4.0
        /// Two adjacent characters swapped.
        public var transposition: Double = 2.0
        /// Floor for any substitution between distinct characters, even when
        /// they share a key position (accent variants reached by long-press:
        /// a→á etc.). Keeps missing-accent typos very cheap but never free.
        public var minSubstitution: Double = 0.35
        /// Cap so that far-apart keys don't produce unbounded costs; a capped
        /// substitution stays comparable to insertion+deletion.
        public var maxSubstitution: Double = 8.0
        /// Substitution involving a character with no key position at all.
        public var unknownCharSubstitution: Double = 5.0
        /// Orthographic (not spatial) confusion pairs: d↔ð, o↔ö, ð↔þ, t↔þ.
        /// These are spatially far on the layout but linguistically common
        /// Icelandic slips, so they get a flat moderate cost.
        public var orthographicConfusion: Double = 1.5
        /// Gaussian width in key widths.
        public var sigma: Double = 0.7
        public init() {}
    }

    /// Icelandic iOS layout rows.
    public static let icelandicRows: [String] = [
        "qwertyuiopð",
        "asdfghjklæö",
        "zxcvbnmþ",
    ]

    /// iOS-style row stagger: home row shifted +0.5, bottom row a further +0.25.
    static let rowOffsets: [Double] = [0.0, 0.5, 0.75]

    /// Horizontal span of the spacebar, in the same key-width units as the
    /// letter grid. On the iOS iPhone layout the bottom function row is
    /// `[123] [globe] [space...] [.] [return]`; the spacebar occupies
    /// roughly the middle five key widths of the ~10-key-wide keyboard.
    /// Bottom-row letters whose key centers fall inside this span sit
    /// directly above the spacebar — a tap intended for space can land on
    /// them (and vice versa). With the current geometry that derives
    /// c v b n m (z, x and þ sit over the 123/globe/./return keys instead).
    static let spacebarXSpan: ClosedRange<Double> = 2.5...7.5

    /// Accented characters are entered via long-press on their base key, so
    /// they share that key's position (the `minSubstitution` floor still
    /// applies between distinct characters).
    static let accentBase: [Character: Character] = [
        "á": "a", "é": "e", "í": "i", "ó": "o", "ú": "u", "ý": "y",
    ]

    /// Common Icelandic orthographic confusions that are NOT spatially close.
    static let confusionPairs: Set<String> = {
        var set = Set<String>()
        for (a, b) in [("d", "ð"), ("o", "ö"), ("ð", "þ"), ("t", "þ")] {
            set.insert(a + b)
            set.insert(b + a)
        }
        return set
    }()

    /// The dedicated physical keys of the layout (the row characters).
    /// Accent twins share their base key's position and are NOT separate
    /// touch targets — the per-tap confidence normalization (see
    /// `PerTapCostProvider`) sums tap likelihood over exactly this set, so
    /// a shared center is counted once.
    static let physicalKeys: [Character] = icelandicRows.flatMap(Array.init)

    public let costs: Costs
    /// Letters on the bottom letter row whose key centers lie within the
    /// spacebar's horizontal span (see `spacebarXSpan`) — derived from the
    /// layout geometry, not hardcoded. A tap on one of these may be a
    /// missed spacebar tap (space-substitution splits in the corrector).
    public let spaceAdjacentLetters: Set<Character>
    private let positions: [Character: SIMD2<Double>]

    /// Key-center of `char` on the unit grid (1.0 = one key pitch in both
    /// axes; y grows downward, row 0 = the q row), nil when the character
    /// has no key position. Accent twins return their base key's center.
    func keyCenter(of char: Character) -> SIMD2<Double>? {
        positions[char]
    }

    public init(costs: Costs = Costs()) {
        self.costs = costs
        var pos: [Character: SIMD2<Double>] = [:]
        for (rowIndex, row) in Self.icelandicRows.enumerated() {
            for (colIndex, char) in row.enumerated() {
                pos[char] = SIMD2(Self.rowOffsets[rowIndex] + Double(colIndex), Double(rowIndex))
            }
        }
        for (accented, base) in Self.accentBase {
            pos[accented] = pos[base]
        }
        self.positions = pos
        let bottomRowIndex = Double(Self.icelandicRows.count - 1)
        self.spaceAdjacentLetters = Set(
            pos.filter { $0.value.y == bottomRowIndex && Self.spacebarXSpan.contains($0.value.x) }
                .keys
        )
    }

    /// -log P(typed | intended) for a single character substitution.
    /// 0 when the characters are equal.
    public func substitutionCost(typed: Character, intended: Character) -> Double {
        if typed == intended { return 0 }
        if Self.confusionPairs.contains(String(typed) + String(intended)) {
            return costs.orthographicConfusion
        }
        guard let p = positions[typed], let q = positions[intended] else {
            return costs.unknownCharSubstitution
        }
        let delta = p - q
        let d2 = delta.x * delta.x + delta.y * delta.y
        let gaussian = d2 / (2 * costs.sigma * costs.sigma)
        return min(max(gaussian, costs.minSubstitution), costs.maxSubstitution)
    }

    /// P(typed char | intended char) as a likelihood in [0, 1].
    public func likelihood(typed: Character, intended: Character) -> Double {
        exp(-substitutionCost(typed: typed, intended: intended))
    }

    /// -log P_spatial(typed | intended): restricted Damerau-Levenshtein
    /// distance where substitutions use key-distance costs and
    /// insert/delete/transpose use the tuned constants.
    public func typingCost(typed: [Character], intended: [Character]) -> Double {
        let n = typed.count
        let m = intended.count
        if n == 0 { return Double(m) * costs.deletion }
        if m == 0 { return Double(n) * costs.insertion }

        // dp[i][j] = cost of producing typed[0..<i] while intending intended[0..<j]
        let width = m + 1
        var dp = [Double](repeating: 0, count: (n + 1) * width)
        for i in 1...n { dp[i * width] = Double(i) * costs.insertion }
        for j in 1...m { dp[j] = Double(j) * costs.deletion }

        for i in 1...n {
            for j in 1...m {
                let sub = dp[(i - 1) * width + (j - 1)]
                    + substitutionCost(typed: typed[i - 1], intended: intended[j - 1])
                let ins = dp[(i - 1) * width + j] + costs.insertion
                let del = dp[i * width + (j - 1)] + costs.deletion
                var best = min(sub, ins, del)
                if i >= 2, j >= 2,
                    typed[i - 1] == intended[j - 2],
                    typed[i - 2] == intended[j - 1],
                    typed[i - 1] != typed[i - 2]
                {
                    best = min(best, dp[(i - 2) * width + (j - 2)] + costs.transposition)
                }
                dp[i * width + j] = best
            }
        }
        return dp[n * width + m]
    }

    /// Convenience overload over strings (assumed lowercased).
    public func typingCost(typed: String, intended: String) -> Double {
        typingCost(typed: Array(typed), intended: Array(intended))
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/SpatialModel.kt
package `is`.solberg.lyklabord.engine

import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

/**
 * Keyboard-adjacency likelihood model for the Icelandic layout.
 *
 * Key centers live on a unit grid (1.0 = one key width) with iOS-style row
 * stagger. Substitution likelihood is Gaussian in the distance between key
 * centers; edit operations use the tuned constants in [Costs]. All costs are
 * in nats.
 */
class SpatialModel(
    val costs: Costs = Costs(),
) {
    class Costs(
        /** User typed an extra character that was not intended. */
        var insertion: Double = 4.0,
        /** User omitted a character they intended. */
        var deletion: Double = 4.0,
        /** Two adjacent characters swapped. */
        var transposition: Double = 2.0,
        /** Floor for a substitution between distinct characters. */
        var minSubstitution: Double = 0.35,
        /** Cap for substitutions between far-apart keys. */
        var maxSubstitution: Double = 8.0,
        /** Substitution involving a character with no key position. */
        var unknownCharSubstitution: Double = 5.0,
        /** Flat cost for common Icelandic orthographic confusions. */
        var orthographicConfusion: Double = 1.5,
        /** Gaussian width in key widths. */
        var sigma: Double = 0.7,
    )

    /** Small JVM replacement for Swift's SIMD2<Double> key-center value. */
    internal data class KeyCenter(val x: Double, val y: Double)

    /** Letters on the bottom row whose centers sit above the spacebar span. */
    val spaceAdjacentLetters: Set<Char>

    private val positions: Map<Char, KeyCenter>

    init {
        val pos = mutableMapOf<Char, KeyCenter>()
        icelandicRows.forEachIndexed { rowIndex, row ->
            row.forEachIndexed { colIndex, char ->
                pos[char] = KeyCenter(rowOffsets[rowIndex] + colIndex.toDouble(), rowIndex.toDouble())
            }
        }
        for ((accented, base) in accentBase) {
            pos[accented] = pos[base]!!
        }
        positions = pos
        val bottomRowIndex = (icelandicRows.size - 1).toDouble()
        spaceAdjacentLetters = pos
            .filter { (_, center) -> center.y == bottomRowIndex && center.x in spacebarXSpan }
            .keys
            .toSet()
    }

    /** Key center of [char], or null when the character has no key position. */
    internal fun keyCenter(of: Char): KeyCenter? = positions[of]

    /** -log P(typed | intended) for one character substitution. */
    fun substitutionCost(typed: Char, intended: Char): Double {
        if (typed == intended) return 0.0
        if (confusionPairs.contains("$typed$intended")) {
            return costs.orthographicConfusion
        }
        val p = positions[typed] ?: return costs.unknownCharSubstitution
        val q = positions[intended] ?: return costs.unknownCharSubstitution
        val deltaX = p.x - q.x
        val deltaY = p.y - q.y
        val d2 = deltaX * deltaX + deltaY * deltaY
        val gaussian = d2 / (2.0 * costs.sigma * costs.sigma)
        return min(max(gaussian, costs.minSubstitution), costs.maxSubstitution)
    }

    /** P(typed | intended) as a likelihood in [0, 1]. */
    fun likelihood(typed: Char, intended: Char): Double =
        exp(-substitutionCost(typed, intended))

    /**
     * Restricted Damerau-Levenshtein typing cost. Substitutions use key-distance
     * costs; insertion, deletion and transposition use [Costs].
     */
    fun typingCost(typed: List<Char>, intended: List<Char>): Double {
        val n = typed.size
        val m = intended.size
        if (n == 0) return m.toDouble() * costs.deletion
        if (m == 0) return n.toDouble() * costs.insertion

        val width = m + 1
        val dp = DoubleArray((n + 1) * width)
        for (i in 1..n) dp[i * width] = i.toDouble() * costs.insertion
        for (j in 1..m) dp[j] = j.toDouble() * costs.deletion

        for (i in 1..n) {
            for (j in 1..m) {
                val sub = dp[(i - 1) * width + (j - 1)] +
                    substitutionCost(typed[i - 1], intended[j - 1])
                val ins = dp[(i - 1) * width + j] + costs.insertion
                val del = dp[i * width + (j - 1)] + costs.deletion
                var best = min(sub, min(ins, del))
                if (
                    i >= 2 && j >= 2 &&
                    typed[i - 1] == intended[j - 2] &&
                    typed[i - 2] == intended[j - 1] &&
                    typed[i - 1] != typed[i - 2]
                ) {
                    best = min(best, dp[(i - 2) * width + (j - 2)] + costs.transposition)
                }
                dp[i * width + j] = best
            }
        }
        return dp[n * width + m]
    }

    /** Convenience overload for lowercased strings. */
    fun typingCost(typed: String, intended: String): Double =
        typingCost(typed.toList(), intended.toList())

    companion object {
        /** Icelandic iOS layout rows. */
        val icelandicRows: List<String> = listOf(
            "qwertyuiopð",
            "asdfghjklæö",
            "zxcvbnmþ",
        )

        /** iOS-style row stagger: home row +0.5, bottom row +0.75. */
        private val rowOffsets: List<Double> = listOf(0.0, 0.5, 0.75)

        /** Horizontal span of the iPhone spacebar in key-width units. */
        private val spacebarXSpan: ClosedRange<Double> = 2.5..7.5

        /** Accented characters share their base key's physical center. */
        val accentBase: Map<Char, Char> = mapOf(
            'á' to 'a', 'é' to 'e', 'í' to 'i', 'ó' to 'o', 'ú' to 'u', 'ý' to 'y',
        )

        /** Inverse of [accentBase]: base vowel to acute twin. */
        val acuteOfBase: Map<Char, Char> = accentBase.entries.associate { (accented, base) ->
            base to accented
        }

        /** Common Icelandic orthographic confusions, directionally including v→ð. */
        internal val confusionPairs: Set<String> = buildSet {
            for ((a, b) in listOf('d' to 'ð', 'o' to 'ö', 'ð' to 'þ', 't' to 'þ')) {
                add("$a$b")
                add("$b$a")
            }
            add("vð")
        }

        /** Directional edge-key undershoot pairs. */
        internal val edgeUndershootPairs: Set<String> = setOf("pð", "læ", "æö", "mþ")

        /** Dedicated physical keys; accent twins are intentionally omitted. */
        internal val physicalKeys: List<Char> = icelandicRows.flatMap { it.toList() }
    }
}
